storage: factor out replicaStateLoader

tbg opened the pull request:
This is one of the self-contained threads to unravel for #18784.

The resulting code allows the `StateLoader` struct to be moved into
its own package (but since moves rot quickly, I'm going to merge this
first and then execute the move separately).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19441)
<!-- Reviewable:end -->

tbg committed:
storage: factor out replicaStateLoader

This is one of the self-contained threads to unravel for #18784.

The resulting code allows the `StateLoader` struct to be moved into
its own package (but since moves rot quickly, I'm going to merge this
first and then execute the move separately).

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 11 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19441#-:-Kx8Y-VS0dQx-fszNAVn:bnfp4nl)*
<!-- Sent from Reviewable.io -->
